### ISSUE TITLE
Disable/Enable specific Rules

### ISSUE THREAD

**Issue by steelbrain** [golang/lint#263]
Hi everyone 👋 

I've performed a github issues search before opening this issue and couldn't find an already existing one. `golint` is great but sometimes you want all the rules except one and it would be wonderful to specify it somewhere. Other modern linters like ESLint support it, would be good to have it supported in golint as well.

**Comment by deitch** [comment]
Been looking for exactly that. E.g. I might not have proper docs on an exported function, but for a given project but might not matter.

**Comment by ericflo** [comment]
I'm in the same boat - don't want to turn off **all** the rules, but the docs on exported functions won't be feasible for a codebase I'm working on.

**Comment by dsnet** [comment]
Hi, I'm on the fence about whether this should be done. In the mean time, I would like to gather more information about why this should be a feature.

My thoughts are:
* The purpose of `lint` is to encourage a consistent style of Go across the community. An individual developer may not agree with its style rules (I don't agree with some of them), but there is a strong benefit to consistency in the Go community. The ability to turn on/off specific rules is a very slippery slope where a person can disable a sufficient number of rules such that the tool can't effectively do its job.
* There are certain rules that lead to many false positives (i.e., it flags something as a style violation that most people would agree is okay). These cases can be annoying, but I would vote that we either tighten the rule or delete these rules than a blanket ability to enable/disable rules.

**Comment by Zanndorin** [comment]
@dsnet I might have some legacy code which I cannot refactor for reasons of time.

But the real reason I'm here is because I wanted to turn off the warning of not commenting public functions in the internal legacy application im coding in.

**Comment by gonzaloserrano** [comment]
I'm with @Zanndorin here, I want to disable the `exported function XXX should have comment or be unexported` check since I'm dealing with legacy code.

**Comment by StephenWeatherford** [comment]
And what if the community has an established code that does not follow some of the rules?  I'm working on an open source project which has about 1500 warnings about not using underlines (and underlines are used because it most closely matches their data model - and they have a confidence of 0.9).  It's unreasonable to expect all communities to have the exact same set of rules in all situations.

**Comment by dominikh** [comment]
It's equally unreasonable to expect golint, which was written for a specific style of Go, to accomodate the needs of all communities that feel it necessary to deviate from that style. 

You're free to fork golint and maintain a version specific to your community.

**Comment by StephenWeatherford** [comment]
@dsnet specifically asked why the feature was needed, and why min_confidence isn't sufficient for particular use cases.  So I'm answering that question as requested.  Having more flexible options would make the tool more useful in general.

**Comment by misham** [comment]
It's helpful to omit a rule in on a single line while preserving it for other cases in a project.  For example, when implementing callback functions for a library I don't need to add documentation to it or I'm dealing with legacy code.

It would be useful to be able to "turn off" a warning for a specific line by adding a comment, similar to how other linters behave.  This way, there is a comment in the code that a lint warning is being ignored and why.   The development team can agree or disagree with this as part of a code review.

By being able to control linter's checks in this manner, I can wire it up to a Makefile or a CI script and have it run on every commit without worrying about false positives or instances where I just don't care enough to fix it.

**Comment by rana** [comment]
Linters are suggestions, not rules. I do observe the linters suggestions, but I sometimes disagree, such as using underscores can be helpful in some cases. Rules are enforced by the compiler and gofmt. Everything else is legal and optional; therefore, it would be nice to disabled some lint suggestions. The challenge is that if i don't conform to every golint suggestion the warnings become too numerous, ignored and I abandon the linter.

**Comment by gaffo** [comment]
Another example I've hit recently is that the linter tells me that fields need to be named in a certain way, like Id should be ID, or no underscores in names.... however I'm working with other systems which have the opposite rules and already specified the protocol and naming so in this case these rules are just generating noise for me. If I was working in an entirely go world then I'd be totally fine with everything following the go rules but I'm working in an existing ecosystem and that's not likely to change. Therefore I have the option of either using the linter and ignoring tons of things with tons of grep -v in my make file and some non-intuitive scripting to see if the linter returned an error that I cared about so I can fail the build... or I can just turn off the linter. I'd rather have a linter that let me use as much as possible of the rules.

**Comment by dominikh** [comment]
Third option: fork the tool, implement the rules of your organisation.

**Comment by gaffo** [comment]
How open would you be to changes that make forks a bit easier but don't implement the feature in the core?

**Comment by gaffo** [comment]
@dominikh ^?

**Comment by dominikh** [comment]
@gaffo that's really a question for @dsnet, the maintainer of golint. But you'd also have to be more specific than "changes". What specific changes would make forks easier?

**Comment by gaffo** [comment]
I was just checking if you guys were antithetical to them. It likely depends on how the main is structured if I can get a list of issues back before they're printed out then it'd be easy to do. if that's all tied up in main I'd likely just want to decouple the main from the error scanner library style.

**Comment by dsnet** [comment]
I may have been the last person to do significant work on this repo, but I don't own this. In fact, I don't think `lint` has a clear owner. For this reason, I don't believe that I am the right person to make calls like this.

That being said, I am ambivalent about this feature. I can see how it helps people since some warnings are really annoying (and my opinion is that they should just be removed), but I see it as a slippery slope also to effectively make lint not enforce anything. In my personal projects, I actually disable the "should have comment or be unexported" rule all the time by just passing the output of `lint` to `grep -v`.

If we were add the ability to disable rules, I like the approach that staticcheck takes, which involves a list of "FILE:RULE" like "github.com/dsnet/compress/brotli/*.go:SA4016", where you have to be explicit about what rule you are disabling and where. However, that presumes that the rules in lint have an ID number, which they don't. We could always add them.

\cc @alandonovan, can we find an actual owner to make decisions on issues like this?

**Comment by gaffo** [comment]
Okay... that's enough willingness to think about it that I'll look into it.

**Comment by gaffo** [comment]
The main does return a list of Problems tho which is what I think I need for a fork I believe

**Comment by jnicholls** [comment]
I agree on the ability to disable certain lints.  Ironically my use case is also the uncommented exported symbols, like a lot of others on this thread.

**Comment by jnicholls** [comment]
On that note, I'm currently instead using `gometalinter` and using the `--exclude` regexp capability to ignore these specific warnings.  This is working well, and may be useful for others on this thread.

**Comment by Zyphrax** [comment]
You should be able to disable certain rules. This is a linter, not a compiler, it helps people to use a common style, but rules shouldn't be set in stone.

I think that by not offering the option to disable rules, people will likely disable the linter when it keeps nagging them, resulting in a lot of bad code.

It is like saying you can't use a screwdriver to open a paint bucket... why not let the user decide.

**Comment by andybons** [comment]
From https://github.com/golang/lint#purpose (emphasis mine):

> The suggestions made by golint are exactly that: suggestions. Golint is not perfect, and has both false positives and false negatives. Do not treat its output as a gold standard. **We will not be adding pragmas or other knobs to suppress specific warnings, so do not expect or require code to be completely "lint-free".** In short, this tool is not, and will never be, trustworthy enough for its suggestions to be enforced automatically, for example as part of a build process.

There are other linters being actively maintained by the community that serve this use case.

**Comment by mikijov** [comment]
I have put together a quick hack for -disable_category and -disable_check options. I am not making it a PR just yet as there does not seem to be willingness for this direction. If there is willingness, I can further improve the code. Please see my fork if you find it useful. Comments are in README.md.
https://github.com/mikijov/lint

**Comment by frankgreco** [comment]
While I completely agree that all go projects should conform to the statement in the purpose of golint, there are use cases where you're using a dep that doesn't conform to it and hence golint will complain even though the change is out of your control.

**Comment by majelbstoat** [comment]
FWIW, a specific use case for this is implementing a GRPC server that has an autogenerated method like `GetUserById` in it, rather than `GetUserByID`. The linter doesn't know the method needs to be called that to conform to the correct interface. And the go protobuf folks have declined to modify the autogenerated code to follow Go naming conventions stating that 'autogenerated code should not be held to the same high standards as human code': https://github.com/golang/protobuf/issues/156

So, yeah, caught between two conflicting world views.

**Comment by fgm** [comment]
Similar need here: on specific cases it is useful to specify the type of a variable, which might be omitted (anonymous functions), but leaving it in place allows easier refactorings by finding all instances of the type, so I would like to have a way to disable linting for the next line.

This seems less intrusive / slippery than disabling a rule entirely. And, with the experience of JS linters, the recent success of Prettier seems to imply keeping configurability to a minimum is a good idea.

**Comment by swhobbit** [comment]
> You're free to fork golint and maintain a version specific to your community.

This is an unfortunate attitude.  

It assumes a community is large enough with extra resources enough to support long term tool chain maintenance.   My community has a size of one: me.   It also assumes forking is free, which untrue: Forking leads to lost changes, be they features or fixes.

No matter what size the community is, forking code should simply, be a last resort.

**Comment by adamrbennett** [comment]
I'll add that the more warnings we have to ignore, the less useful the linter becomes. The example given above of 1500 linter warnings is a good example. Under this scenario, a new warning will easily be missed and the linter has failed to do its job -- which is not to simply show the warnings, but to encourage proper style, which is dependent upon the visibility and prominence of warnings. Too many warnings decreases visibility.

I understand the motivation to resist a lint config. It allows for too broad of a deviation from go style standards. However, I think line item disabling via comments would be a nice compromise.

**Comment by tnm-** [comment]
@adamrbennett Exactly. I enabled lint, got over a 100 warning, a handful of them was useful. I wanted to automatically ignore the rest. Since it's not possible, I'm gonna disable lint as it is now 95+% noise.
I think it would have been OK, to add the most simple workaround, e.g. adding a special comment like //lintskip to an "offending" line.

**Comment by borud** [comment]
I think this is a principle versus practicality discussion.  I would prefer to run the linter continually while I develop to catch any mistakes, omissions and typos.  This is useful as it flags problems as you are writing code and not later.  This isn't practical if the IDE constantly flags code as problematic when it isn't.  I could change the code, but quite frankly, in some cases you do want to be able to have non-public types with public functions.  Of course I could skip every file where the linter throws up suggestions, but then what would be the point?

The tool simply doesn't work for my regular test-vet-lint-build cycle and I've stopped using it on most projects.  I don't understand what is so important about not implementing a "nolint" comment or an option that allows you to filter out certain rules when so many users appear to have the same problem.

And it isn't a simple matter of forking the linter or using a different one.  I'm not interested in spending time wondering what linter to use.  I'm not interested in having to re-check every now and then if the linter is being maintained.

And no, the -min_confidence flag is not an attractice solution because then you burden the user with the task of having to figure out what other problems are NOT being listed.  And perhaps you do want to see some of those suggestions that get filtered out.

**Comment by shellscape** [comment]
The ratio of reactions here would suggest the maintainer(s) are vastly outnumbered and disagreed with by the the community the tool is supposed to support.

**Comment by Waitak** [comment]
For me, the issue here is that some suggestions are based on incorrect assumptions that the linter makes. The "stutter" warning is a case in point for us. I have the following types:

```
type SCACQueryParams struct { ... }
type SCACsQueryParams struct { ... }

```
These are used to call functions called GetSCAC() and GetSCACs(), respectively. The first generates the warning:

```
type name will be used as scac.SCACQueryParams by other packages, and that stutters; consider calling this QueryParams.
```

You can see the thought behind the warning, but it doesn't apply here. The suggestion, in fact, would be a _terrible_ idea. I'd love to be able to tell the linter "Okay, I see your point, but there's a good reason why that doesn't apply here. Stand down on this one."

**Comment by RavenHursT** [comment]
Wow.. the intellectual pride here is staggering.  To assume that _every_ engineer and organization is going to want to write the _exact_ same style of code is obtuse to say the least.  This is why I loathe "linting" in general.  If the code compiles, then it compiles.  People need to quit trying to force others to conform to their subjective views of writing code and instead learn how to adapt to reading code that isn't exactly the way one would want to see code written.

As a result, I too will be disabling this tool.

**Comment by sebnyberg** [comment]
I agree with the majority here that there should be an option to turn off certain linting rules when they don't make sense - at least locally in a file. 

As mentioned before, the Protobuf / gRPC generated types already break the standard, but some editors (such as VSCode) has built in a "hack" that checks for the code generation comment to disable warnings:

```
// Code generated by protoc-gen-go. DO NOT EDIT.
```

It's a terrible solution, but I prefer putting this magic comment in my files over introducing a tool that is not built into the language.

**Comment by hndada** [comment]
Ironically, forcing a user follows all rules in golint make the user *not to use* it at all, which causes advance inconsistency of go codes in overall community.

**Comment by jacobisaliveandwell** [comment]
> Ironically, forcing a user follows all rules in golint make the user _not to use_ it at all, which causes advance inconsistency of go codes in overall community.

This point is paramount. All hail `gofmt`, but a good portion of the linting is best practices, and a simple tool isn't smart enough to know the trade offs a developer may have considered.

Today I needed to fork a Terraform Provider. The repo has a very common pattern that go-lint complains about, `should not use basic type string as key in context.WithValue`, except Terraform probably has a good reason for why it does. Even if it doesn't, there isn't anything I can't do anything about it.

Please reconsider this. Anything that doesn't break the compiler should be configurable.

**Comment by iddan** [comment]
My workaround:
```bash
#!/bin/bash

# Lint all the files using go lint excluding errors which are currently explicitly ignore
# This script is intended to be used in the continuous integration process
# When editing, it is highly recommended to use ShellCheck (https://www.shellcheck.net/) to avoid common pitfalls

# Patterns to be ignored from the go lint output
IGNORED_PATTERNS=(
    " comment "
)

# Patterns joined into a regular expression
REGEX=$(printf "|(%s)" "${IGNORED_PATTERNS[@]}")
REGEX=${REGEX:1}

# Execute go lint on all the files and filter output by the regualr expression
output=$( ( (golint ./... | egrep -v "$REGEX") 2>&1 ) | tee /dev/fd/2);
if [ -z "$output" ]
then
    exit 0
else
    exit 1
fi
```

**Comment by dgoodine** [comment]
My workaround: removed golint from our CI pipeline.

@dsnet's wariness about switches and knobs makes a fair point. Which is all the reason why _rules_ need to be more clearly though out before they're established. The rule that using "ID" and not "Id" for an identifier is beyond the pale of neurotic style enforcement.

Tools that are otherwise useful but go beyond practical reasoning become more trouble than they're worth. And people will stop using them.

**Comment by RavenHursT** [comment]
@dgoodine Inevitable end to "linting"... #justLetFolksCode

**Comment by ghostsquad** [comment]
So it sounds like it's more important for the Go team to have a "all or nothing" stance than it is to make a useful tool?

I get that it's certainly possible (and currently being done) for the community to create their own tools. That's only saying that the standard library (of tools) is less adequate than what the community creates to replace them. You might as well not even create the tools if you don't care about real world usage.

The [Zen of Python](https://www.python.org/dev/peps/pep-0020/) puts this nicely:

> Special cases aren't special enough to break the rules.
> Although practicality beats purity.

[Pep 8](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds) also has some wisdom

> A Foolish Consistency is the Hobgoblin of Little Minds

> A style guide is about consistency. Consistency with this style guide is important. Consistency within a project is more important. Consistency within one module or function is the most important.

> However, know when to be inconsistent -- sometimes style guide recommendations just aren't applicable. When in doubt, use your best judgment. Look at other examples and decide what looks best. And don't hesitate to ask!

> In particular: do not break backwards compatibility just to comply with this PEP!

> Some other good reasons to ignore a particular guideline:

> 1. When applying the guideline would make the code less readable, even for someone who is used to reading code that follows this PEP.
> 2. To be consistent with surrounding code that also breaks it (maybe for historic reasons) -- although this is also an opportunity to clean up someone else's mess (in true XP style).
> 3. Because the code in question predates the introduction of the guideline and there is no other reason to be modifying that code.
> 4. When the code needs to remain compatible with older versions of Python that don't support the feature recommended by the style guide.

---

Specifically let me call out "use your best judgment". Accept that maybe _your_ way of doing things doesn't work for everyone.

It's quite unfortunate that a programming language, which has a literal purpose allowing people to create things in the world that don't exist yet, or to modify the world around them, has such strict requirements around how _it_ should be used. Even if you know better, it's also OK to let people learn and even fail.

**Comment by jcollum** [comment]
> So it sounds like it's more important for the Go team to have a "all or nothing" stance than it is to make a useful tool?

The Go ecosystem is usually a "do it the Go way or go away" kind of thing.

**Comment by ianlancetaylor** [comment]
The many Go developers have tried to develop a framework for tooling that can be used by tools like lint.  There is also a specific lint tool that uses that tooling and tests various style guidelines, as outlined at https://go.googlesource.com/lint/+/refs/heads/master/README.md.  If you want a different tool that meets different goals, you can use use the tooling framework to write your own tool.

To put it another way, there is one tool that the Go developers feel should be used by everyone.  That tool is `go vet`.  There is also this lint tool.  Nobody thinks that this tool should be used by everyone.  It's just a convenient tool for some people, and an example tool for people who want something similar but not exactly the same.

**Comment by ghostsquad** [comment]
I just find it a bit ridiculous that practical reasons, of which there are many described in this thread, are dismissed as unnecessary knobs. That's like telling a aircraft pilot, that his control plane has too many knobs, and that he should use an iPhone (with just one button) instead.

**Comment by ianlancetaylor** [comment]
That's a good analogy.  I would say that the lint tool is an iPhone, and you want an aircraft.  There is a role for both iPhones and aircraft in this world.  We can help you build your aircraft if you want.  It just won't be the lint tool.

(There are several other lint tools available already, and perhaps one of them is more to your liking.  For example, you might want to look at https://pkg.go.dev/honnef.co/go/tools/staticcheck?tab=doc).

**Comment by jcollum** [comment]
Thanks for giving us the lay of the land there @ianlancetaylor

**Comment by jacobisaliveandwell** [comment]
I get that tools not run through `go` (e.g. `go vet`) are not considered first class citizens, but I thought this discussion process was the means by which tools become productionalized. Closing this issue is effectively saying this is a pet project and nothing more.

**Comment by ianlancetaylor** [comment]
No, this is a way of saying that this specific project has specific goals, and is not intended to be all things to all people.

**Comment by ghostsquad** [comment]
@ianlancetaylor we aren't asking it to be all things to all people. We are asking to be able to fill the fuel tank ourselves, instead of requiring someone else to fill it. We are asking for Visual Flight Rules & Instrumental Flight Rules (piloting), because the _real world_ has weather and low visibility.

We are not asking for it to be complicated. Simple is good. But without knobs, it's like driving a car without:

1. Radio (silence is superior)
2. A/C (just roll down your window)
3. GPS (just remember the route and plan before hand)

If you feel strongly about the rules of linting, make them default, but allow them to be configurable. That's all anyone here is asking for.

Anyways... I don't think this is productive anymore. Frankly, this type of thread just says a lot about how the team _doesn't_ believe that `Practicality beats Purity`.

**Comment by ianlancetaylor** [comment]
Use the underlying tooling to do whatever you like.  Or just copy the source code of this project and change it however you like.  Perhaps your new project will become more popular than this one, and we can retire this one.

This project has a defined set of goals, as mentioned above.  There is nothing wrong with having different goals.  The Go project wants to give you the tools to meet those goals.  Where we haven't given you those tools, please let us know.

But this particular project is meant to support a specific defined set of goals.  It is not meant to be a general purpose program.  It is not meant to be used by people who do not share the goals of this program.

Perhaps we just need to rename this project.  Perhaps people see "lint" and think that it is something they should use.  For Go, the thing you should use it called "vet".  You should only be using this project, "lint", if you think it has good goals.  If you don't, you should be using a different project.  There are many tools out there for linting and checking Go code.  There is nothing special about this one.

**Comment by jacobisaliveandwell** [comment]
Honestly, I just read this as you asserting your beliefs. This doesn't match anything I've ever heard from talks by core contributors.

**Comment by ianlancetaylor** [comment]
It's not clear to me why you think that I am arguing against discussion.  There is plenty of discussion here.  I haven't said that it should stop.

A decision was made on this issue back in https://github.com/golang/lint/issues/263#issuecomment-363652697.  More discussion here isn't going to change that decision.  But by all means continue to discuss.

I'm commenting here because I see people expressing unhappiness and dissatisfaction with the Go project, and I'm trying to explain why the decision is what it is, and I'm trying to point out the other options that people have to achieve their goals.

**Comment by ghostsquad** [comment]
> The Go project wants to give you the tools to meet those goals. Where we haven't given you those tools, please let us know.

See this thread...

> Perhaps we just need to rename this project. Perhaps people see "lint" and think that it is something they should use. For Go, the thing you should use it called "vet". You should only be using this project, "lint", if you think it has good goals. If you don't, you should be using a different project. There are many tools out there for linting and checking Go code. There is nothing special about this one.

Perhaps. I think the problem here is that once a tool is made "by the official team", people tend to latch on to it, and prefer it over other homegrown tools. This is just my perception. Sometimes those "official tools" don't meet the standards of and requirements of some, and community built tools do become more popular. Let's look at history to see how that pans out:

1. pkg/errors -> Go Team realized that errors are too simple, and some additional knobs are needed
2. dependency management -> deb, glide, etc are invented -> Go Team realizes that "just use the HEAD of master" is not as practical, and some additional knobs are needed, and we land on go modules.
3. generics -> cheekybits/genny is invented, because requiring every developer to write code generation themselves is actually really burdensome -> hopefully a generics proposal is accepted soon and the _compiler_ can do the code generation.

Please don't misinterpret this as me being against "start simple and iterate". Indeed, I am all for starting simple and iterating.

**Comment by adamrbennett** [comment]
If I write a tool for my personal use and the community doesn't like it I would respond in the same way the maintainers have responded here.

If I write a tool to support a community and it doesn't fit their reasonable needs, I would alter it.

Is this tool built for the community or for the maintainers?

On the topic of project goals: is one of the goals of this project to support the reasonable needs of the community?

**Comment by ghostsquad** [comment]
> Is this tool built for the community or for the maintainers?

> On the topic of project goals: is one of the goals of this project to support the reasonable needs of the community?

I definitely think these are super relevant question here.

**Comment by ianlancetaylor** [comment]
Those are good questions.

I have opened a proposal at https://golang.org/issue/38968 to simply remove the lint program.  See that issue for my reasoning.

I encourage people to comment on that issue, preferably about the meta-topic of whether the lint program should exist or not rather than details of what the lint program should do.

Thanks for the comments.

**Comment by ghostsquad** [comment]
> I have opened a proposal at golang.org/issue/38968 to simply remove the lint program. See that issue for my reasoning.

Upvoted

**Comment by dgoodine** [comment]
Upvoted. Thank you @ianlancetaylor.

**Comment by phanirithvij** [comment]
It's same with the [dartfmt](https://github.com/dart-lang/dart_style/issues/534) maintainers another great framework (Flutter) built for the community, by our beloved Google

**Comment by patarapolw** [comment]
A use case is, not-yet-fixed golint issues...

https://github.com/golang/lint/issues/393

**Comment by tpiekarski** [comment]
> * There are certain rules that lead to many false positives (e.g., it flags something as a style violation that most people would agree is okay). These cases can be annoying, but I would vote that we either tighten the rule or delete these rules than a blanket ability to enable/disable rules.

My fifty cents and supporting what @dsnet wrote. Take the rule about "vars declared but not used" for example.

When I am coding and writing a new variable this is **immediately highlighted as an error** in IDEs such Visual Studio Code. Perhaps it could be possible to deactivate it there, but I did not found any way so far. Do you?

Well this is **annoying and distracting**. At this moment it does not do anything good for clean code.

Don't get me wrong I like to use linters, they should be used and are very important. 
**But I think as soon as something gets annoying people tend to stop using it.**

**Please add a way to deactivate rules.**
